File: sdk/python/intentprotocol/sanitize.py
"""Sanitization for display-safe content (v0.2 anti-phishing)."""

from __future__ import annotations

import re
from typing import Any

URL_PATTERN = re.compile(
    r"https?://\S+|www\.\S+|\b[\w.-]+\.(com|fr|org|net|io)\b",
    re.IGNORECASE,
)
PHONE_PATTERN = re.compile(r"\d[\d\s.\-]{6,}\d")


def sanitize_for_display(s: str | None) -> str:
    """Remove URLs and mask phone-like sequences for safe user display."""
    if s is None or not isinstance(s, str):
        return ""
    out = URL_PATTERN.sub("[url removed]", s)
    out = PHONE_PATTERN.sub("[phone]", out)
    return out


def validate_display_field(s: str | None) -> tuple[bool, str | None]:
    """Check if string contains URL or phone pattern. Returns (ok, reason)."""
    if s is None or not isinstance(s, str):
        return True, None
    if URL_PATTERN.search(s):
        return False, "URL not allowed"
    if PHONE_PATTERN.search(s):
        return False, "Phone pattern not allowed"
    return True, None


def sanitize_bid_for_display(bid: dict[str, Any]) -> dict[str, Any]:
    """Return a copy of bid with offer display fields sanitized."""
    if not bid or "offer" not in bid:
        return bid
    offer = dict(bid["offer"])
    if "location" in offer and offer["location"]:
        loc = dict(offer["location"])
        if isinstance(loc.get("name"), str):
            loc["name"] = sanitize_for_display(loc["name"])
        if isinstance(loc.get("address"), str):
            loc["address"] = sanitize_for_display(loc["address"])
        offer["location"] = loc
    if isinstance(offer.get("service"), str):
        offer["service"] = sanitize_for_display(offer["service"])
    return {**bid, "offer": offer}

File: sdk/js/src/sanitize.js
/**
 * Sanitization for display-safe content (v0.2 anti-phishing).
 * Strip URLs and mask or remove phone-like patterns from text shown to users.
 */

const URL_PATTERN = /https?:\/\/[^\s]+|www\.[^\s]+|\b[\w.-]+\.(com|fr|org|net|io)\b/gi;
const PHONE_PATTERN = /\d[\d\s.\-]{6,}\d/g;

/**
 * Sanitize a string for safe display: remove URLs, mask phone-like sequences.
 * @param {string} str - Raw string (e.g. from location.name, offer.service)
 * @returns {string} Sanitized string safe for user display
 */
export function sanitizeForDisplay(str) {
  if (str == null || typeof str !== 'string') return '';
  let out = str.replace(URL_PATTERN, '[url removed]');
  out = out.replace(PHONE_PATTERN, (m) => '[phone]');
  return out;
}

/**
 * Check if a string contains URL or phone pattern (for validation before sending).
 * @param {string} str
 * @returns {{ ok: boolean, reason?: string }}
 */
export function validateDisplayField(str) {
  if (str == null || typeof str !== 'string') return { ok: true };
  if (str.match(URL_PATTERN)) return { ok: false, reason: 'URL not allowed' };
  if (str.match(PHONE_PATTERN)) return { ok: false, reason: 'Phone pattern not allowed' };
  return { ok: true };
}

/**
 * Sanitize a bid's offer for display (location.name, location.address, service).
 * Use before rendering bid content to users.
 * @param {import('./types.js').BidMessage} bid
 * @returns {Object} Copy of bid with offer display fields sanitized
 */
export function sanitizeBidForDisplay(bid) {
  if (!bid?.offer) return bid;
  const offer = { ...bid.offer };
  if (offer.location) {
    offer.location = { ...offer.location };
    if (typeof offer.location.name === 'string') offer.location.name = sanitizeForDisplay(offer.location.name);
    if (typeof offer.location.address === 'string') offer.location.address = sanitizeForDisplay(offer.location.address);
  }
  if (typeof offer.service === 'string') offer.service = sanitizeForDisplay(offer.service);
  return { ...bid, offer };
}
